### ISSUE TITLE
Provide a convenient way for a developer to build the release variant or all variants of a project

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle-native#140]
Currently, a developer can build the release variant using the `linkRelease` task for libraries or the `installRelease` task for executables. It would be useful to offer a consistent lifecycle task that doesn't require the developer, or tooling, to know what kind of output the project produces.

To build both variants, the developer can run these tasks plus the `assemble` task. Again, it would be useful to offer a lifecycle tasks that produces all the variants of the project without the developer or tooling knowing what kind of output is produced.

Personally, I don't think we should change the `assemble` task to build all variants. Building just enough for the developer to do something useful with the project is better for this task as well as the `check` and `build` tasks.

**Comment by eljobe** [comment]
@adammurdoch, have you considered the possibility of taking variant specification as a kind of command-line flag which influences which tasks participate in a specific lifecycle rather than having to add new lifecycle tasks? For example, we might image `gradle --variants release,debug assemble` I am nervous about adding additional lifecycle tasks to Gradle’s model for some variants (e.g. release and debug) but not offering a generic mechanism for additional variants the user may come up with. (e.g. chocolate and pistachio)

**Comment by adammurdoch** [comment]
Yes, there were a bunch of options in the design doc that we might implement. They're not all mutually exclusive.

It certainly wouldn't be the case that 'debug' and 'release' would be treated any differently to any custom variants. So, if we opt for adding an `assembleRelease` task, it would be implemented in such a way that any custom variant X would automatically get an `assembleX` task defined.

A downside of using command-line options - or any new CLI behaviour - is that it is not particularly usable or discoverable from a clickity-click UI (I can't just select something from a list of task), is not discoverable from the CLI, and is completely unusable from build logic to implement custom workflows or orchestration or aggregations or whatever. Or at least without us doing a bunch of work to introduce new concepts.

A downside of using tasks to define the groupings is that you need a task for every grouping x action.

A good solution would eventually sit in the middle, I think. There are super common groupings (specifically, "build me something I can test", "build me something I can release", "build me everything that you can on this machine") that should be easy to use out of the box and are important to be reachable from build logic for orchestration. There could and even should be lifecycle tasks for these, but we might look at other options, such as APIs to query a 'buildable' for a particular combination that the command-line, the tapi and build logic can all drive as they need to.

This story, however, should be scoped to the simplest increment that adds value.

**Comment by lacasseio** [comment]
I agree with the discussion so far. One issue I had with the software model implementation was the order in which the dimension appeared in the task name. It was always a pain to know what comes before/after when typing a task name. The workflow looks mostly like this:
- `./gradlew tasks`
- Find and copy the name of the task you want to invoke
- `./gradlew <past the name of the task to invoke>`

This workflow happens several times a day and it was quite annoying. The issue with a CLI flag to filter, which can be easily implemented by the build author, is when you want to assemble a custom split as a single invocation: Build debug x86 and release x64. I can't say this happened a lot but it was a major step up compared to other build tools. You can go grab a coffee while this is happening. You can also trust Gradle to do the right thing.
